File: sbr ios/SBR/SBR/Models/AgentInventory.swift
import Foundation

struct AgentInventoryItem: Codable, Identifiable {
    var id: String { _id ?? UUID().uuidString }
    let _id: String?
    let agentId: String?
    let posProductId: Int?
    let productId: String?
    let productName: String
    let sku: String?
    let category: String?
    let quantity: Int
    let minThreshold: Int?
    let unitPrice: Double?
    let lastUpdated: String?
    let createdAt: String?
    let updatedAt: String?
    
    var isLowStock: Bool {
        return quantity <= (minThreshold ?? 1)
    }
    
    var displayName: String {
        return productName
    }
    
    enum CodingKeys: String, CodingKey {
        case _id, id, agentId, posProductId, productId, productName, name, sku, category, quantity, minThreshold, minAlertThreshold, unitPrice, price, lastUpdated, createdAt, updatedAt
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.container(keyedBy: CodingKeys.self)
        self._id = (try? container.decodeIfPresent(String.self, forKey: ._id)) ?? (try? container.decodeIfPresent(String.self, forKey: .id))
        
        // agentId can be string or populated user object
        if let agentStr = try? container.decodeIfPresent(String.self, forKey: .agentId) {
            self.agentId = agentStr
        } else {
            self.agentId = nil
        }
        
        self.posProductId = try? container.decodeIfPresent(Int.self, forKey: .posProductId)
        
        // productId can be string id or null
        if let prodStr = try? container.decodeIfPresent(String.self, forKey: .productId) {
            self.productId = prodStr
        } else {
            self.productId = nil
        }
        
        let pName = (try? container.decodeIfPresent(String.self, forKey: .productName))
            ?? (try? container.decodeIfPresent(String.self, forKey: .name))
            ?? "Component / Part"
        self.productName = pName
        
        self.sku = try? container.decodeIfPresent(String.self, forKey: .sku)
        self.category = try? container.decodeIfPresent(String.self, forKey: .category)
        self.quantity = (try? container.decodeIfPresent(Int.self, forKey: .quantity)) ?? 0
        self.minThreshold = (try? container.decodeIfPresent(Int.self, forKey: .minThreshold))
            ?? (try? container.decodeIfPresent(Int.self, forKey: .minAlertThreshold))
            ?? 1
        self.unitPrice = (try? container.decodeIfPresent(Double.self, forKey: .unitPrice))
            ?? (try? container.decodeIfPresent(Double.self, forKey: .price))
        self.lastUpdated = try? container.decodeIfPresent(String.self, forKey: .lastUpdated)
        self.createdAt = try? container.decodeIfPresent(String.self, forKey: .createdAt)
        self.updatedAt = try? container.decodeIfPresent(String.self, forKey: .updatedAt)
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        try container.encodeIfPresent(_id, forKey: ._id)
        try container.encodeIfPresent(agentId, forKey: .agentId)
        try container.encodeIfPresent(posProductId, forKey: .posProductId)
        try container.encodeIfPresent(productId, forKey: .productId)
        try container.encode(productName, forKey: .productName)
        try container.encodeIfPresent(sku, forKey: .sku)
        try container.encodeIfPresent(category, forKey: .category)
        try container.encode(quantity, forKey: .quantity)
        try container.encodeIfPresent(minThreshold, forKey: .minThreshold)
        try container.encodeIfPresent(unitPrice, forKey: .unitPrice)
        try container.encodeIfPresent(lastUpdated, forKey: .lastUpdated)
        try container.encodeIfPresent(createdAt, forKey: .createdAt)
        try container.encodeIfPresent(updatedAt, forKey: .updatedAt)
    }
}

File: sbr android/app/src/main/java/com/sbr/sms/data/models/AgentInventory.kt
package com.sbr.sms.data.models

import com.google.gson.annotations.SerializedName

enum class IndentStatus(val displayName: String) {
    REQUESTED("Pending Store Review"),
    PENDING("Pending Store Review"),
    APPROVED("Approved"),
    DISPATCHED("Dispatched to Van"),
    REJECTED("Rejected")
}

data class ProductRef(
    @SerializedName("id", alternate = ["_id"]) val id: String = "",
    val name: String = "",
    val sku: String? = null,
    val price: Double? = 0.0,
    val stockLevel: Int? = 0,
    val minStockLevel: Int? = 0
) {
    val _id: String get() = id
}

data class AgentInventoryItem(
    @SerializedName("id", alternate = ["_id"]) val id: String = "",
    val agentId: Any? = null,
    val posProductId: Int? = null,
    val productId: Any? = null,
    @SerializedName("productName", alternate = ["name"]) val productName: String = "Spare Part",
    val sku: String? = null,
    val category: String? = "General Spares",
    val quantity: Int = 0,
    @SerializedName("minThreshold", alternate = ["minAlertThreshold"]) val minThreshold: Int? = 1,
    @SerializedName("unitPrice", alternate = ["price"]) val unitPrice: Double? = 0.0,
    val lastUpdated: String? = null,
    val lastRestockedAt: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val _id: String get() = id
    val minAlertThreshold: Int get() = minThreshold ?: 1
    val isLowStock: Boolean get() = quantity <= (minThreshold ?: 1)
    val displayName: String get() = productName.takeIf { it.isNotBlank() } ?: "Spare Part"

    val resolvedProductId: String
        get() {
            if (productId is String && productId.isNotBlank()) return productId
            if (productId is Map<*, *>) {
                val idVal = productId["id"] ?: productId["_id"]
                if (idVal is String) return idVal
            }
            return id
        }
}

data class IndentItem(
    val posProductId: Int? = null,
    val productId: Any? = null,
    @SerializedName("name", alternate = ["productName"]) val name: String = "Spare Part",
    val sku: String? = null,
    @SerializedName("requestedQuantity", alternate = ["quantity"]) val requestedQuantity: Int = 1,
    val dispatchedQuantity: Int? = null
) {
    val resolvedProductId: String
        get() {
            if (productId is String) return productId
            if (productId is Map<*, *>) {
                val idVal = productId["id"] ?: productId["_id"]
                if (idVal is String) return idVal
            }
            return ""
        }
}

data class AgentIndent(
    @SerializedName("id", alternate = ["_id"]) val id: String = "",
    val agentId: HandoverUser? = null,
    @SerializedName("serviceRequestId", alternate = ["requestId"]) val serviceRequestId: String? = null,
    val items: List<IndentItem> = emptyList(),
    val status: String = "PENDING", // PENDING, REQUESTED, APPROVED, DISPATCHED, REJECTED
    val urgency: String? = "MEDIUM", // LOW, MEDIUM, HIGH, CRITICAL
    val agentRemarks: String? = null,
    val inchargeRemarks: String? = null,
    @SerializedName("storeInchargeId", alternate = ["dispatchedBy"]) val storeInchargeId: HandoverUser? = null,
    val requestedAt: String? = null,
    val dispatchedAt: String? = null,
    val createdAt: String? = null,
    val updatedAt: String? = null
) {
    val _id: String get() = id
}
